### ISSUE TITLE
[BUG - Release] Error: [Lua::Registry::get_function_ref] Ref was not function No traceback

### ISSUE THREAD

**Issue by BurgersMcFly** [UE4SS-RE/RE-UE4SS#346]
**Branch or Release**
Latest version 3.0, but it was happening with previous experimental release as well. D:

**Game and Engine Version**
Palworld, 5.1.1

**Describe the bug**
Hooking on certain functions across multiple mods causes this error when the hook initiates: 
Error: [Lua::Registry::get_function_ref] Ref was not function No traceback

This makes some hooks not function, essentially making some mods not functional.

Moving the hooks in the same file fixes this issue.

The mods have enabled.txt in them, but setting the problematic one (the one that isn't functioning because of the error) to enabled in mods.txt fixes the issue but you still get the error. (????????) *Correction: Doing this actually makes the hook from DisplayDateTime not function, so I guess all it did was change which hook actually hooked.

**To Reproduce**
[1]Download my mod https://www.nexusmods.com/palworld/mods/348?tab=files&file_id=1257
[2]Download my mod https://www.nexusmods.com/palworld/mods/501 version 1.0.0.0, 
[3]Install mods 
[4]Launch Palworl
[5]Load into your worl
[6]Open map
[7]Get error
[8]FastTravelFromAnywhere is not working

**Expected behavior**
Expected for the hooks to work properly? Like why is this affecting only some functions? Have found two till now.

**Screenshots and UE4SS Log**
There's nothing else in the log, everything is working fine until you trigger the hook and get that error. Nothing else. D:

**Desktop (please complete the following information):**
 - OS: Win10

**Additional context**
I've tried moving the code to functions and calling the functions in the hooks. I've tried hooking the hooks with a delay so that they are hooked at different times. I've tried removing the ExecuteAsync or adding it to DisplayDateTime but nothing works. D:

Relevant code from DisplayDateTime:

```lua
RegisterHook("/Game/Pal/Blueprint/UI/UserInterface/Map/WBP_Map_Base.WBP_Map_Base_C:OnSetup", function(Context)
		if not customized then
			local fontMaterial = AssignFontMaterial(config.color)
			local textCursorLocation = Context:get().Text_CursorLocation
	
			textCursorLocation.Font["Size"] = config.fontSize
			textCursorLocation.Font["TypefaceFontName"] = FName(config.fontType)
			textCursorLocation.Font["FontMaterial"] = fontMaterial
	
			print("[DisplayDateTime] Did DisplayDateTime customizations!")
		end
		customized = true
	end)
```

Relevant code from FastTravelFromAnywhere:	
	
```lua
RegisterHook("/Game/Pal/Blueprint/UI/UserInterface/Map/WBP_Map_Base.WBP_Map_Base_C:OnSetup", function(Context)
		Context:get():SetPropertyValue("Can Fast Travel", true)
		ExecuteAsync(function()
			for _, PalLocationPointFastTravel in ipairs(PalLocationPointFastTravels) do
				PalLocationPointFastTravel.ShouldUnlockFlag = false
			end
		end)
	end)
```

**Comment by UE4SS** [comment]
Looking at your scripts, I noticed that you're registering the same hooks multiple times by calling `RegisterHook` in other hooks.
The only reason I can think of to do that is if the UFunction stops existing at some point, which usually isn't something that happens.
Are you sure you're supposed to be doing that ?

Regarding `ExecuteAsync`, avoid this function unless you truly need your code to be executed asynchronously.

As far as your error is concerned, I'm not sure what's causing it.
It's only produced if there's no Lua function stored in the Lua registry for your hook.
This error is not an error I would expect from misusing our API.
This will need to be further investigated.

**Comment by UE4SS** [comment]
I'm unable to reproduce this problem in my test game, and I don't have Palworld.

**Comment by BurgersMcFly** [comment]
"I noticed that you're registering the same hooks multiple times by calling RegisterHook in other hooks."

Can you explain to me where this is? I'd like to fix this. But I'm pretty sure the hooks initiate only after ClientRestart or ServerAcknowledge and will only hook once depending of which of the two happens first. They are wrapped inside a function that gets called in ClientRestart and ServerAcknowledge that both have a check so that they don't do the hooks ever again if hooked once.

"Regarding ExecuteAsync, avoid this function unless you truly need your code to be executed asynchronously."

Yea coming from Hogwarts, ExecuteAsync and ExecuteWithDelay triggers ptsd LOL. But I feel like ExecuteAsync helps with microstutters when doing the for loop.

"I'm unable to reproduce this problem in my test game, and I don't have Palworld."

Yea I wouldn't expect this to be replicatable(word?) outside of Palworld and the specific functions that cause it, since it seems to happen only with two until now.

**Comment by UE4SS** [comment]
> "I noticed that you're registering the same hooks multiple times by calling RegisterHook in other hooks."
> 
> Can you explain to me where this is? I'd like to fix this. But I'm pretty sure the hooks initiate only after ClientRestart or ServerAcknowledge and will only hook once depending on which of the two happens first. They are wrapped inside a function that gets called in ClientRestart and ServerAcknowledge that both have a check so that they don't do the hooks ever again if hooked once.

The `ClientRestart` function isn't guaranteed to only be executed once, it's usually executed when a player spawns in, and some games will end up triggering `ClientRestart` upon a player respawning after dying, not just the initial spawn when loading the game world, but even if it was just when first loading the game world, games usually have main menus and if you were to load the world, then go back to the main menu and load the world again then `ClientRestart` will have executed at least twice and maybe more depending on the game devs.

I haven't looked into `ServerAcknowledge` but the name suggests that it might trigger every time a player connects to the co-op server, which sounds like it also might happen more than once.

**Comment by UE4SS** [comment]
If you're properly checking if it's been hooked before hooking in `ClientRestart`, it should be fine (though not hot-reload safe so beware when developing the mod) but I just wanted to say that it's generally not recommended, and I've seen a bunch of mods do all sorts of horrific infinite chain hooking.

**Comment by BurgersMcFly** [comment]
Yea ClientRestart and ServerAcknowledge do trigger again on mount/dismount, just like Hogwarts lol, which is why I've added the checks so that the functions that contain the hooks are actually execute only once. Or whenever it is needed for the server (if player ~= lastplayer).

Thank you for your explanation, time and help.

I decided to just use a different hook for DisplayDateTime, but this is still a problem with my other mods, since I'm using a hook to revert the lastPlayer to nil when returning to the main menu (prevents crashing when re-entering worlds). Since this seems to be function specific I will try to find another function to hook to in order to do this, but I felt you guys had to know since it felt like a bug.

**Comment by UE4SS** [comment]
Could you try this experimental build ? https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/zDEV-UE4SS_v3.0.0-1-g6a0f235.zip
I took a quick look at the code and noticed that we were passing the wrong Lua instance for non-native functions and the `OnSetup` function looks like it might be non-native based on the path.

**Comment by BurgersMcFly** [comment]
Still the same error when the hook is supposed to be executed D:

**Comment by UE4SS** [comment]
I decided to do a performance review of your DisplayDateTime mod.

These next two notes are regarding the `WBP_Map_Base_C:Tick` hook.

Major Problem: Call to `StaticFindObject` on every tick is slow because looking up an object in GUObjectArray is not particularly fast.
Solution: Cache `/Script/Pal.Default__PalUtility`.
Likely safe to cache without cache invalidation checks as it's a CDO, which means it's not likely to ever be destroyed.

Minor Problem: String comparison in a tick function. String comparisons are generally a lot slower than numeric (bools included) comparisons.
I consider his a minor problem because the string is very short, and I don't know whether some string optimizations exists for very short strings.
Solution: Remove `am_pm`, and make a new variable to hold `true/false`, let's call it `is_am`, depending on the hour as determined by the already existing if-statement.
Then replace `am_pm` in the format with `(is_am and "AM" or "PM")`.

Other Problem: I don't know if `WBP_Map_Base_C:BP_GetDesiredFocusTarget` is a hot-path but I'll mention this just in case it is: You're always calling `StaticFindObject` when retrieving fonts, you might've already done this but it's worth mentioning anyways, just in case, that you should look into whether the game or UE ever unloads these fonts, because if they never get unloaded then you should be able to store the font objects themselves instead of their names which will give you better performance.
Of course, even if they do get unloaded, you could create & manage a cache for them similar to what you do with the player object, and in that case, see `NotifyOnNewObject` (in our docs) since `ClientRestart` is not appropriate for fonts.
Obviously, this isn't really a problem at all unless that particular hook gets hit a lot, or if users are complaining about performance when they trigger whatever action that causes this hook to fire.

**Comment by UE4SS** [comment]
> Still the same error when the hook is supposed to be executed D:
> 
> Edit: Non-native meaning bp generated? Because it is bp generated.
> 
> Edit2: At least I think that's how it's called.

Yes.
Native means the function is machine code, which is what C++ will generate when it compiles the function.
Non-native, means it's anything else, and in the case of UE that means it's a BP function, which are executed by a VM.

**Comment by BurgersMcFly** [comment]
**Major Problem: Call to StaticFindObject on every tick**

I know D: It hurts to publish code like this but, I don't know if it's Palworld, UE5 or people's PCs, when I originally posted my mods with cached Default objects, people where reporting nul references. I asked in the Palworld modding discord server, since this was a first to me as well, and they said that the object is either getting destroyed or whatever else is going on. I guess I could cache it in the beginning of the script and add a if object nul then StaticFind, but that would add an additional if in the Tick function D:

**Minor Problem: String comparison in a tick function**

Hm.. didn't know this could be a problem, tbh I'm running rivatuner statistics while developing with the frame frametime history graph enabled to keep an eye out for performance degradation. I can't say I've noticed anything, but you're def right with doing all this in a Tick function. But it was the only way to implement this unfortunately and I tried to keep it as simple as possible, which is why I do the font customization stuff in another hook.

**Other Problem**

The code inside the hook is executed only once, when you first open the Map, so I don't think it's going to cause any issues? I've changed the hook again in a later update. I originally used NotifyOnNewObject but it made it extremely complicated and just went with a hook instead.

Thank you so much for the info and help. I will def take a look at your suggested solution for the Minor Problem. 

**ClientRestart is not appropriate for fonts**

But ClientRestart  is used to check if the player has loaded into the world and then executes the function for the hooks to be registered. The font customization code is executed only after the Map is opened.

**Comment by UE4SS** [comment]
> **Major Problem: Call to StaticFindObject on every tick**
> 
> It hurts to publish code like this but, I don't know if it's Palworld, UE5 or people's PCs, when I originally posted my mods with cached Default objects, people where reporting nul references. I asked in the Palworld modding discord server, since this was a first to me as well, and they said that the object is either getting destroyed or whatever else is going on. I guess I could cache it in the beginning of the script and add an if object nul then StaticFind, but that would add an additional if in the Tick function D:
> 
> **Minor Problem: String comparison in a tick function**
> 
> Hm.. didn't know this could be a problem, tbh I'm running rivatuner statistics while developing with the frametime history overlay graph enabled to keep an eye out for performance degradation. I can't say I've noticed anything, but you're def right with doing all this in a Tick function. But it was the only way to implement this unfortunately and I tried to keep it as simple as possible, which is why I do the font customization stuff in another hook.
> 
> **Other Problem**
> 
> The code inside the hook is executed only once, when you first open the Map, so I don't think it's going to cause any issues? I've changed the hook again in a later update. I originally used NotifyOnNewObject but it made it extremely complicated and just went with a hook instead.
> 
> **ClientRestart is not appropriate for fonts**
> 
> But ClientRestart is used to check if the player has loaded into the world and then executes the function for the hooks to be registered. The font customization code is executed only after the Map is opened and it only happens once. After that the flag is set to true to prevent it from running again.
> 
> Thank you so much for the info and help. I will def take a look at your suggested solution for the Minor Problem.


People say you shouldn't optimize if you can't see a noticeable performance impact and I usually agree with that but not when it comes to code that runs on every tick.
I personally think that code running on every tick should be as optimized as possible because small stuff adds up and even more so in these types of functions.
What happens is, you'll profile your code and not see much of a problem, and then the code changes a bunch down the line (especially problematic if other people change the code as they might not be as strict with profiling as you) and eventually you end up with ten performance problems to fix instead of just one problem, and obviously that will be much more overwhelming to fix, not to mention all the comments you'll get from your users if the problem gets bad enough to be noticeable in gameplay.
Basically, when I see something obviously less performant with an easy fix in a tick function, I tend to prefer fixing it, even if it won't be noticeable at the moment.
But nevertheless, it's likely fine as it is, which is why I labeled it as a minor problem.

The `ClientRestart` function is not suitable for fonts simply because it has nothing to do with when the fonts get created, for that you need to either hook a function that gets called on font creation or use `NotifyOnNewObject`.
And the font problem is just something that I noticed and wanted to make you aware of because then you could make an informed decision based on your better knowledge of this particular hook regarding if it should be optimized, and it sounds like it shouldn't be optimized.

**Comment by BurgersMcFly** [comment]
Yea I def agree with you. Sorry if it seemed like it didn't. I'm just saying that I'm not sure if I can do much apart from your suggested fix for the minor issue.

**Comment by UE4SS** [comment]
> Yea I def agree with you. Sorry if it seemed like it didn't. I'm just saying that I'm not sure if I can do much apart from your suggested fix for the minor issue.

Not at all, just trying to convey my message as clearly as I can.

**Comment by BurgersMcFly** [comment]
Gotcha, thanks again for the help! :D

**Comment by UE4SS** [comment]
I've discovered what's causing this problem and I'm working on a fix.

**Comment by BurgersMcFly** [comment]
That's great! :D

**Comment by UE4SS** [comment]
Should be fixed in #351 and here's the experimental build: https://github.com/UE4SS-RE/RE-UE4SS/releases/download/experimental/zDEV-UE4SS_v3.0.0-1-g7a0a941.zip
Please let me know if this fixed it for you.

**Comment by BurgersMcFly** [comment]
Yep!! It worked! Hooks are functional with no errors and both mods are working fine! Good job! :D

Btw I've updated the DisplayDateTime mod and I've given you credits in the description. Let me know if you want me to credit you under a different name!

**Comment by UE4SS** [comment]
> Yep!! It worked! Hooks are functional with no errors and both mods are working fine! Good job! :D
> 
> Btw I've updated the DisplayDateTime mod and I've given you credits in the description. Let me know if you want me to credit you under a different name!
> 
> Edit: Is this going to be released under the release branch?

Probably just link directly to the comment that helped you, that way it can maybe help others as well.

I've made a pull request, it now needs to be approved by someone else, and if it's approved it will be merged into main and eventually make into a release.
Approval could take a few days, we have a limited number of people and we're all very busy.

**Comment by BurgersMcFly** [comment]
Yep that's a great idea, didn't know I can link directly to comments!

I see. It's okay, thank you for looking into this and fixing it so fast and thank you for all the tips and help. And I almost forgot.. Congrats on releasing v3! :D
